### PR TITLE
fixes #8534 : :Add i18-next for this "Please reschedule."

### LINKED ISSUE #8534: :Add i18-next for this
https://api.github.com/calcom/cal.com/blob/1eeb91a793e577b10c3037a2158d7ad052659b20/packages/trpc/server/routers/viewer/bookings/requestReschedule.handler.ts#L232

```typescript

import type { BookingReference, EventType, WebhookTriggerEvents } from "@prisma/client";
import { BookingStatus, WorkflowMethods } from "@prisma/client";
import type { TFunction } from "next-i18next";

import { getCalendar } from "@calcom/app-store/_utils/getCalendar";
import { cancelScheduledJobs } from "@calcom/app-store/zapier/lib/nodeScheduler";
import { CalendarEventBuilder } from "@calcom/core/builders/CalendarEvent/builder";
import { CalendarEventDirector } from "@calcom/core/builders/CalendarEvent/director";
import { deleteMeeting } from "@calcom/core/videoClient";
import dayjs from "@calcom/dayjs";
import { deleteScheduledEmailReminder } from "@calcom/ee/workflows/lib/reminders/emailReminderManager";
import { deleteScheduledSMSReminder } from "@calcom/ee/workflows/lib/reminders/smsReminderManager";
import { sendRequestRescheduleEmail } from "@calcom/emails";
import { getCalEventResponses } from "@calcom/features/bookings/lib/getCalEventResponses";
import getWebhooks from "@calcom/features/webhooks/lib/getWebhooks";
import sendPayload from "@calcom/features/webhooks/lib/sendPayload";
import { isPrismaObjOrUndefined } from "@calcom/lib";
import { getTranslation } from "@calcom/lib/server";
import { prisma } from "@calcom/prisma";
import type { CalendarEvent, Person } from "@calcom/types/Calendar";

import { TRPCError } from "@trpc/server";

import type { TrpcSessionUser } from "../../../trpc";
import type { TRequestRescheduleInputSchema } from "./requestReschedule.schema";
import type { PersonAttendeeCommonFields } from "./types";

type RequestRescheduleOptions = {
  ctx: {
    user: NonNullable<TrpcSessionUser>;
  };
  input: TRequestRescheduleInputSchema;
};

export const requestRescheduleHandler = async ({ ctx, input }: RequestRescheduleOptions) => {
  const { user } = ctx;
  const { bookingId, rescheduleReason: cancellationReason } = input;

  const bookingToReschedule = await prisma.booking.findFirstOrThrow({
    select: {
      id: true,
      uid: true,
      userId: true,
      title: true,
      description: true,
      startTime: true,
      endTime: true,
      eventTypeId: true,
      eventType: true,
      location: true,
      attendees: true,
      references: true,
      customInputs: true,
      dynamicEventSlugRef: true,
      dynamicGroupSlugRef: true,
      destinationCalendar: true,
      smsReminderNumber: true,
      scheduledJobs: true,
      workflowReminders: true,
      responses: true,
    },
    where: {
      uid: bookingId,
      NOT: {
        status: {
          in: [BookingStatus.CANCELLED, BookingStatus.REJECTED],
        },
      },
    },
  });

  if (!bookingToReschedule.userId) {
    throw new TRPCError({ code: "FORBIDDEN", message: "Booking to reschedule doesn't have an owner" });
  }

  if (!bookingToReschedule.eventType) {
    throw new TRPCError({ code: "FORBIDDEN", message: "EventType not found for current booking." });
  }

  const bookingBelongsToTeam = !!bookingToReschedule.eventType?.teamId;

  const userTeams = await prisma.user.findUniqueOrThrow({
    where: {
      id: user.id,
    },
    select: {
      teams: true,
    },
  });

  if (bookingBelongsToTeam && bookingToReschedule.eventType?.teamId) {
    const userTeamIds = userTeams.teams.map((item) => item.teamId);
    if (userTeamIds.indexOf(bookingToReschedule?.eventType?.teamId) === -1) {
      throw new TRPCError({ code: "FORBIDDEN", message: "User isn't a member on the team" });
    }
  }
  if (!bookingBelongsToTeam && bookingToReschedule.userId !== user.id) {
    throw new TRPCError({ code: "FORBIDDEN", message: "User isn't owner of the current booking" });
  }

  if (bookingToReschedule) {
    let event: Partial<EventType> = {};
    if (bookingToReschedule.eventTypeId) {
      event = await prisma.eventType.findFirstOrThrow({
        select: {
          title: true,
          users: true,
          schedulingType: true,
          recurringEvent: true,
        },
        where: {
          id: bookingToReschedule.eventTypeId,
        },
      });
    }
    await prisma.booking.update({
      where: {
        id: bookingToReschedule.id,
      },
      data: {
        rescheduled: true,
        cancellationReason,
        status: BookingStatus.CANCELLED,
        updatedAt: dayjs().toISOString(),
      },
    });

    // delete scheduled jobs of previous booking
    cancelScheduledJobs(bookingToReschedule);

    //cancel workflow reminders of previous booking
    bookingToReschedule.workflowReminders.forEach((reminder) => {
      if (reminder.method === WorkflowMethods.EMAIL) {
        deleteScheduledEmailReminder(reminder.id, reminder.referenceId);
      } else if (reminder.method === WorkflowMethods.SMS) {
        deleteScheduledSMSReminder(reminder.id, reminder.referenceId);
      }
    });

    const [mainAttendee] = bookingToReschedule.attendees;
    // @NOTE: Should we assume attendees language?
    const tAttendees = await getTranslation(mainAttendee.locale ?? "en", "common");
    const usersToPeopleType = (
      users: PersonAttendeeCommonFields[],
      selectedLanguage: TFunction
    ): Person[] => {
      return users?.map((user) => {
        return {
          email: user.email || "",
          name: user.name || "",
          username: user?.username || "",
          language: { translate: selectedLanguage, locale: user.locale || "en" },
          timeZone: user?.timeZone,
        };
      });
    };

    const userTranslation = await getTranslation(user.locale ?? "en", "common");
    const [userAsPeopleType] = usersToPeopleType([user], userTranslation);

    const builder = new CalendarEventBuilder();
    builder.init({
      title: bookingToReschedule.title,
      type: event && event.title ? event.title : bookingToReschedule.title,
      startTime: bookingToReschedule.startTime.toISOString(),
      endTime: bookingToReschedule.endTime.toISOString(),
      attendees: usersToPeopleType(
        // username field doesn't exists on attendee but could be in the future
        bookingToReschedule.attendees as unknown as PersonAttendeeCommonFields[],
        tAttendees
      ),
      organizer: userAsPeopleType,
    });

    const director = new CalendarEventDirector();
    director.setBuilder(builder);
    director.setExistingBooking(bookingToReschedule);
    cancellationReason && director.setCancellationReason(cancellationReason);
    if (event) {
      await director.buildForRescheduleEmail();
    } else {
      await director.buildWithoutEventTypeForRescheduleEmail();
    }

    // Handling calendar and videos cancellation
    // This can set previous time as available, until virtual calendar is done
    const credentialsMap = new Map();
    user.credentials.forEach((credential) => {
      credentialsMap.set(credential.type, credential);
    });
    const bookingRefsFiltered: BookingReference[] = bookingToReschedule.references.filter((ref) =>
      credentialsMap.has(ref.type)
    );
    bookingRefsFiltered.forEach(async (bookingRef) => {
      if (bookingRef.uid) {
        if (bookingRef.type.endsWith("_calendar")) {
          const calendar = await getCalendar(credentialsMap.get(bookingRef.type));

          return calendar?.deleteEvent(bookingRef.uid, builder.calendarEvent, bookingRef.externalCalendarId);
        } else if (bookingRef.type.endsWith("_video")) {
          return deleteMeeting(credentialsMap.get(bookingRef.type), bookingRef.uid);
        }
      }
    });

    // Send emails
    await sendRequestRescheduleEmail(builder.calendarEvent, {
      rescheduleLink: builder.rescheduleLink,
    });

    const evt: CalendarEvent = {
      title: bookingToReschedule?.title,
      type: event && event.title ? event.title : bookingToReschedule.title,
      description: bookingToReschedule?.description || "",
      customInputs: isPrismaObjOrUndefined(bookingToReschedule.customInputs),
      ...getCalEventResponses({
        booking: bookingToReschedule,
        bookingFields: bookingToReschedule.eventType?.bookingFields ?? null,
      }),
      startTime: bookingToReschedule?.startTime ? dayjs(bookingToReschedule.startTime).format() : "",
      endTime: bookingToReschedule?.endTime ? dayjs(bookingToReschedule.endTime).format() : "",
      organizer: userAsPeopleType,
      attendees: usersToPeopleType(
        // username field doesn't exists on attendee but could be in the future
        bookingToReschedule.attendees as unknown as PersonAttendeeCommonFields[],
        tAttendees
      ),
      uid: bookingToReschedule?.uid,
      location: bookingToReschedule?.location,
      destinationCalendar:
        bookingToReschedule?.destinationCalendar || bookingToReschedule?.destinationCalendar,
      cancellationReason: `Please reschedule. ${cancellationReason}`, // TODO::Add i18-next for this
    };

    // Send webhook
    const eventTrigger: WebhookTriggerEvents = "BOOKING_CANCELLED";
    // Send Webhook call if hooked to BOOKING.CANCELLED
    const subscriberOptions = {
      userId: bookingToReschedule.userId,
      eventTypeId: (bookingToReschedule.eventTypeId as number) || 0,
      triggerEvent: eventTrigger,
    };
    const webhooks = await getWebhooks(subscriberOptions);
    const promises = webhooks.map((webhook) =>
      sendPayload(webhook.secret, eventTrigger, new Date().toISOString(), webhook, {
        ...evt,
        smsReminderNumber: bookingToReschedule.smsReminderNumber || undefined,
      }).catch((e) => {
        console.error(`Error executing webhook for event: ${eventTrigger}, URL: ${webhook.subscriberUrl}`, e);
      })
    );
    await Promise.all(promises);
  }
};

```

### PR DESCRIPTION
**Pull request by kuluruvineeth**
## What does this PR do?

- Added i18-next for this "Please reschedule."
- Before: cancellationReason: `Please reschedule. ${cancellationReason}`
- After: cancellationReason: t('Please reschedule.') + " " + cancellationReason

Fixes #8534

**Environment**: Production

### CODE REVIEW

**Review by hariombalhara — CHANGES_REQUESTED** [review]
Thanks for the contribution @kuluruvineeth 🙏
But we shouldn't concatenate translated strings
Screenshot 2023-05-02 at 6 40 40 PM

We should instead use complete string with variable  `Please reschedule. {{cancellationReason}}`

See e.g. https://github.com/calcom/cal.com/blob/f060e8e543edb78606284b2129ec4ca6f304ec30/apps/web/public/static/locales/en/common.json#L34

Also, I think you translated it in al the languages yourself. You don't need to do this. We use crowdin to do the translation.

**Comment by kuluruvineeth** [comment]
> Thanks for the contribution @kuluruvineeth 🙏 But we shouldn't concatenate translated strings Screenshot 2023-05-02 at 6 40 40 PM
> 
> We should instead use complete string with variable `Please reschedule. {{cancellationReason}}`
> 
> See e.g.
> 
> https://github.com/calcom/cal.com/blob/f060e8e543edb78606284b2129ec4ca6f304ec30/apps/web/public/static/locales/en/common.json#L34
> 
> Also, I think you translated it in al the languages yourself. You don't need to do this. We use crowdin to do the translation.

@hariombalhara thank you reviewing and suggesting changes, i have made the changes as recommended by passing variable to translated string and added only in en file. Please review and let me know if still any changes are to be made, thanks.

File: packages/trpc/server/routers/viewer/bookings/requestReschedule.handler.ts [hunk 1]
```diff
@@ -229,7 +231,7 @@ export const requestRescheduleHandler = async ({ ctx, input }: RequestReschedule
       location: bookingToReschedule?.location,
       destinationCalendar:
         bookingToReschedule?.destinationCalendar || bookingToReschedule?.destinationCalendar,
-      cancellationReason: `Please reschedule. ${cancellationReason}`, // TODO::Add i18-next for this
+      cancellationReason: t("Please reschedule.", { cancellationReasonMessage: cancellationReason })
```
Comment (emrysal) [inline packages/trpc/server/routers/viewer/bookings/requestReschedule.handler.ts:234]:
```suggestion
      cancellationReason: t("request_reschedule_prompt", { cancellationReasonMessage: cancellationReason })
```

File: apps/web/public/static/locales/en/common.json [hunk 2]
```diff
@@ -1800,5 +1800,6 @@
   "payment_app_commission": "Require payment ({{paymentFeePercentage}}% + {{fee, currency}} commission per transaction)",
   "email_invite_team": "{{email}} has been invited",
   "error_collecting_card": "Error collecting card",
-  "image_size_limit_exceed": "Uploaded image shouldn't exceed 5mb size limit"
+  "image_size_limit_exceed": "Uploaded image shouldn't exceed 5mb size limit",
+  "Please reschedule.": "Please reschedule. {{cancellationReasonMessage}}"
```
Comment (emrysal) [inline apps/web/public/static/locales/en/common.json:1804]:
```suggestion
  "request_reschedule_prompt": "Please reschedule. {{cancellationReasonMessage}}"
```

**Review by emrysal — CHANGES_REQUESTED** [review]
Consistency on the underscoring of keys 🙏

**Comment by kuluruvineeth** [comment]
> Consistency on the underscoring of keys 🙏

@emrysal Thanks for reviewing it. I have renamed to the format of underscoring keys to maintain consistency as suggested. Thanks.

**Comment by joeauyeung** [comment]
I'm closing this for now, as the cancellation reason is already included in the email that is sent.

image

**Pull request closed without merging**